### PR TITLE
feat: e2e test harness with daemon + telnet

### PR DESCRIPTION
**Pull request by abesto**
In an integration test, spin up `telnet-host` and `daemon` with a minimal DB core (that's `Minimal.db` + `eval`). Test a round-trip of `; 42`.

Points of interest:
* Each `#[test]` gets its own brand-new `daemon` and `telnet-host` process. Overhead of this seems negligible at the moment.
* The started processes are guaranteed to be killed via `run_test_as` + `Drop`.
* The runtime test database dir of each `daemon` is a `tempfile::TempDir` that also gets automatically cleaned up.
* Output of `eval` is sandwiched between `-=!-v-!=-` and `-=!-^-!=-`, making output capture & verification straightforward.
* Output of the `daemon` and `telnet-host` subprocesses respect `cargo test -- --nocapture`, that is: they're hidden by default in passing tests.

Possible next steps:
* Nothing! This is enough to serve as a smoke test, and we should leave more full-featured integration tests as they currently are. There's no sense in going through all this machinery for every single test.
* Migrate existing integration tests to this approach. There's no sense in maintaining two different ways of running integration tests; we should maintain a single version, and that should be the higher-fidelity one. (This is the approach ToastStunt took; do with that what you will :P)

### CODE REVIEW

File: crates/daemon/tests/common/mod.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            .current_release()
+            .run()
+            .expect("Failed to build moor-telnet-host")
+            .path()
+            .to_owned()
+    })
+}
+
+// Because of this function, it might be cleaner to host integration tests
+// in a separate crate.
```
Comment (abesto) [inline crates/daemon/tests/common/mod.rs:79]: ... or actually, maybe `telnet-host` would be a better place? /shrug

**Comment by abesto** [comment]
Lol. Test job is failing with:

```
[daemon]: Message:  Unable to create pager. You may need to set /proc/sys/vm/overcommit_memory to '1': InitializationError("Mmap failed for size class block_size: 4096, virt_size 1099511627776: Cannot allocate memory (os error 12)")
```

Hopefully we can tell the container to set overcommit :D

**Comment by rdaum** [comment]
> Lol. Test job is failing with:
> 
> ```
> [daemon]: Message:  Unable to create pager. You may need to set /proc/sys/vm/overcommit_memory to '1': InitializationError("Mmap failed for size class block_size: 4096, virt_size 1099511627776: Cannot allocate memory (os error 12)")
> ```
> 
> Hopefully we can tell the container to set overcommit :D

Three thoughts here:

- We should be able to start this daemon with a lower heap size which would prevent this. (Like the other integration tests), and that should be possible via CLI. Right now it's hardcoded in DatabaseBuilder and shouldn't be.
- The storage engine should have options for not using the Umbra-style vm overcommit buffer pool manager and just using some other buffer allocation (https://github.com/rdaum/moor/issues/294)
- My work on integration the wiredtiger storage engine as an alternative or replacement for my own home-grown would also eliminate this problem

File: crates/daemon/Cargo.toml [hunk 2]
```diff
@@ -39,3 +39,10 @@ ed25519-dalek.workspace = true
 pem.workspace = true
 rand.workspace = true
 rusty_paseto.workspace = true
+
+# Testing
+[dev-dependencies]
+assert_cmd = "2.0.14"
```
Comment (rdaum) [inline crates/daemon/Cargo.toml:45]: We follow a convention of having dependency versions declared in the root workspace Cargo.toml, and then use .workspace = true style dependencies in the workspace crates. This way we're attempting to enforce one version of deps for all crates without risk of drift. Even for dev deps.

File: crates/daemon/tests/common/mod.rs (hunk 1 above)
Comment (rdaum) [reply crates/daemon/tests/common/mod.rs:79]: Yeah, I think the `tests/` directory of `telnet-host` is the right place for that. 

For modularity/coupling reasons in fact I think there should be no references into `telnet-host` from `daemon` at all. So it'd be a matter of moving this code into `crates/telnet-host/tests` having a `dev-dependency` inside `telnet-host/Cargo.toml` referring to the daemon? 

In general that's the approach integration tests should be following for cargo project (see integration tests here) though I have personally not been consistent with that.

https://doc.rust-lang.org/book/ch11-03-test-organization.html

File: crates/telnet-host/src/main.rs [hunk 3]
```diff
@@ -47,6 +47,9 @@ struct Args {
         default_value = "tcp://0.0.0.0:7898"
     )]
     narrative_server: String,
+
+    #[arg(long, help = "Enable debug logging", default_value = "false")]
```
Comment (rdaum) [inline crates/telnet-host/src/main.rs:51]: Didn't this change already get made in another commit?  Need a rebase?

**Review by rdaum — CHANGES_REQUESTED** [review]
Looks great, thanks so much! 

Approved modulo a few nit-changes

**Comment by rdaum** [comment]
> > Lol. Test job is failing with:
> > ```
> > [daemon]: Message:  Unable to create pager. You may need to set /proc/sys/vm/overcommit_memory to '1': InitializationError("Mmap failed for size class block_size: 4096, virt_size 1099511627776: Cannot allocate memory (os error 12)")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hopefully we can tell the container to set overcommit :D
> 
> Three thoughts here:
> 
>     * We should be able to start this daemon with a lower heap size which would prevent this. (Like the other integration tests), and that should be possible via CLI. Right now it's hardcoded in DatabaseBuilder and shouldn't be.
> 
>     * The storage engine should have options for not using the Umbra-style vm overcommit buffer pool manager and just using some other buffer allocation ([Provide non-vm overcommit-based buffer pool allocator option #294](https://github.com/rdaum/moor/issues/294))
> 
>     * My work on integration the wiredtiger storage engine as an alternative or replacement for my own home-grown would also eliminate this problem

I just added a `max_buffer_pool_bytes` CLI arg, you should try using that.

File: crates/daemon/Cargo.toml (hunk 2 above)
Comment (abesto) [reply crates/daemon/Cargo.toml:45]: Ah right. Will do.

File: crates/telnet-host/src/main.rs (hunk 3 above)
Comment (abesto) [reply crates/telnet-host/src/main.rs:51]: Negative, that was `daemon`, this is `telnet-host`.

File: crates/telnet-host/tests/common/mod.rs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    ManagedChild::new(
+        "daemon",
+        Command::cargo_bin("moor-daemon")
+            .expect("Failed to find moor-daemon binary")
+            .arg("--textdump")
+            .arg(minimal_db)
+            .arg("--generate-keypair")
+            .arg("--max-buffer-pool-bytes")
+            .arg((1 << 24).to_string())
```
Comment (rdaum) [inline crates/telnet-host/tests/common/mod.rs:57]: Maybe toss this value into a constant with a comment

File: Cargo.toml
[the remark this answers is not in the document]
Comment (rdaum) [reply Cargo.toml:1]: Eh, I can live with it.  Hopefully my tools (RustRover, emacs) don't disagree with your tools on what the right formatting here is.

**Review by rdaum — APPROVED** [review]

File: crates/telnet-host/tests/common/mod.rs (hunk 4 above)
Comment (abesto) [reply crates/telnet-host/tests/common/mod.rs]: Done!

**Pull request merged**